Python/bytecodes.c

```diff
@@ -148,12 +148,24 @@ dummy_func(
         };
 
         op(_CHECK_PERIODIC, (--)) {
-            CHECK_EVAL_BREAKER();
+            _Py_CHECK_EMSCRIPTEN_SIGNALS_PERIODICALLY();
+            QSBR_QUIESCENT_STATE(tstate); \
+            if (_Py_atomic_load_uintptr_relaxed(&tstate->eval_breaker) & _PY_EVAL_EVENTS_MASK) {
+                if (_Py_HandlePending(tstate) != 0) {
+                    GOTO_ERROR(error); \
+                }
+            }
         }
 
         op(_CHECK_PERIODIC_NOT_YIELD_FROM, (--)) {
             if ((oparg & RESUME_OPARG_LOCATION_MASK) < RESUME_AFTER_YIELD_FROM) {
-                CHECK_EVAL_BREAKER();
+                _Py_CHECK_EMSCRIPTEN_SIGNALS_PERIODICALLY();
+                QSBR_QUIESCENT_STATE(tstate); \
+                if (_Py_atomic_load_uintptr_relaxed(&tstate->eval_breaker) & _PY_EVAL_EVENTS_MASK) {
+                    if (_Py_HandlePending(tstate) != 0) {
+                        GOTO_ERROR(error); \
+                    }
+                }
```

Comment:
Removing the macro is the point of this PR. Macros are opaque to the code generator, so we need to special case them (which is OK for something very common like `Py_DECREF`). Avoiding the special case simplifies the code generator considerably.
It is the call to `_Py_HandlePending` that escapes, not the whole of `_CHECK_PERIODIC` and we want the code generator to know that.

Reply:
I meant a macro instruction of the DSL.

Reply:
Ah, I see.
No it can't, as macros can only be composed a linear sequence. We need branching here, not deoptimization.
`EXIT_IF` and `DEOPT_IF` only make sense in specialized instructions.